src/Categories/Tactic/Category.agda

```diff
@@ -0,0 +1,184 @@
+
+open import Categories.Category
+
+module Categories.Tactic.Category where
+
+open import Level
+open import Function using (_⟨_⟩_)
+
+open import Data.Bool    as Bool    using (Bool; _∨_; if_then_else_)
+open import Data.Maybe   as Maybe   using (Maybe; just; nothing; maybe)
+open import Data.List    as List    using (List; _∷_; [])
+open import Data.Product as Product using (_×_; _,_)
+
+open import Agda.Builtin.Reflection
+open import Reflection.Argument
+open import Reflection.Term using (getName; _⋯⟅∷⟆_)
+open import Reflection.TypeChecking.Monad.Syntax
+
+module _ {o ℓ e} (𝒞 : Category o ℓ e) where
+
+  open Category 𝒞
+  open HomReasoning
+  open Equiv
+
+  private
+    variable
+      A B C : Obj
+      f g   : A ⇒ B
+
+  --------------------------------------------------------------------------------
+  -- An 'Expr' reifies the parentheses/identity morphisms of some series of 
+  -- compositions of morphisms into a data structure. In fact, this is also
+  -- a category!
+  --------------------------------------------------------------------------------
+  data Expr : Obj → Obj → Set (o ⊔ ℓ) where
+    _∘′_ : ∀ {A B C} → Expr B C → Expr A B → Expr A C 
+    id′  : ∀ {A} → Expr A A
+    [_↑] : ∀ {A B} → A ⇒ B → Expr A B
+  
+  -- Embed a morphism in 'Expr' back into '𝒞' without normalizing.
+  [_↓] : Expr A B → A ⇒ B 
+  [ f ∘′ g ↓] = [ f ↓] ∘ [ g ↓]
+  [ id′ ↓]    = id
+  [ [ f ↑] ↓] = f
+  
+  -- Convert an 'Expr' back into a morphism, while normalizing
+  --
+  -- This actually embeds the morphism into the category of copresheaves
+  -- on 𝒞, which obeys the category laws up to beta-eta equality.
+  -- This lets us normalize away all the associations/identity morphisms.
+  embed : Expr B C → A ⇒ B → A ⇒ C
+  embed (f ∘′ g) h  = embed f (embed g h)
+  embed id′ h       = h
+  embed [ f ↑] h    = f ∘ h
+
+  
+  preserves-≈′ : ∀ (f : Expr B C) → (h : A ⇒ B) → embed f id ∘ h ≈ embed f h
+  preserves-≈′ id′ f      = identityˡ
+  preserves-≈′ [ x ↑] f   = ∘-resp-≈ˡ identityʳ
+  preserves-≈′ (f ∘′ g) h = begin
+    embed (f ∘′ g) id ∘ h         ≡⟨⟩
+    embed f (embed g id) ∘ h      ≈˘⟨ preserves-≈′ f (embed g id) ⟩∘⟨refl ⟩
+    (embed f id ∘ embed g id) ∘ h ≈⟨ assoc  ⟩
+    embed f id ∘ embed g id ∘ h   ≈⟨ refl⟩∘⟨ preserves-≈′ g h ⟩
+    embed f id ∘ embed g h        ≈⟨ preserves-≈′ f (embed g h) ⟩
+    embed (f ∘′ g) h              ∎
+  
+  preserves-≈ : ∀ (f : Expr A B) → embed f id ≈ [ f ↓]
+  preserves-≈ id′      = refl
+  preserves-≈ [ x ↑]   = identityʳ
+  preserves-≈ (f ∘′ g) = begin
+    embed (f ∘′ g) id       ≈˘⟨ preserves-≈′ f (embed g id) ⟩
+    embed f id ∘ embed g id ≈⟨ preserves-≈ f ⟩∘⟨ preserves-≈ g ⟩
+    [ f ↓] ∘ [ g ↓]         ≡⟨⟩
+    [ f ∘′ g ↓]             ∎
+
+--------------------------------------------------------------------------------
+-- Reflection Helpers
+--------------------------------------------------------------------------------
+
+_==_ = primQNameEquality
+{-# INLINE _==_ #-}
```

Comment:
I am wondering whether stdlib has provided basic definitions like this?

Reply:
Apparently [not](https://github.com/agda/agda-stdlib/search?q=primQNameEquality). Might be good to add it to `Reflection.Name`?

Reply:
that would be reasonable.

Reply:
Yeah, that would be a good call. I stole this part shamelessly from the `MonoidSolver` in the stdlib, so if that gets added, we ought to fix that as well.